### PR TITLE
planner: fix wrong result when enabling dynamic mode in partition table for tiflash

### PR DESCRIPTION
**Pull request by wshwsh12**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/37254

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Fix wrong result when enabling dynamic mode in partition table for tiflash
```

### CODE REVIEW

File: planner/core/fragment.go [hunk 1]
```diff
@@ -301,7 +301,17 @@ func partitionPruning(ctx sessionctx.Context, tbl table.PartitionedTable, conds
...
 	}
 	if len(ret) == 0 {
-		ret = []table.PhysicalTable{tbl.GetPartition(pi.Definitions[0].ID)}
+		// TiFlash cannot process an empty task correctly, so choose to leave it with some data to read.
+		if len(partitionNames) == 0 {
+			ret = []table.PhysicalTable{tbl.GetPartition(pi.Definitions[0].ID)}
+		} else {
+			for i, def := range pi.Definitions {
+				if def.Name.L == partitionNames[0].L {
+					ret = []table.PhysicalTable{tbl.GetPartition(pi.Definitions[i].ID)}
```
Comment (windtalker) [inline planner/core/fragment.go:310]: why not just use `ret = []table.PhysicalTable{tbl.GetPartition(def.ID)}` here

Comment (wshwsh12) [reply planner/core/fragment.go]: Addressed.

**Review by windtalker — APPROVED** [review]
LGTM

**Review by XuHuaiyu — APPROVED** [review]

**Pull request merged**